sui_storage/src/lock_service.rs

```diff
@@ -0,0 +1,491 @@
+                        pending_transaction: previous_tx_digest,
+                    });
+                }
+            } else {
+                // Only write the locks that need to be written (are uninitialized)
+                let obj_ref = owned_input_objects[i];
+                locks_to_write.push((obj_ref, Some(tx_digest)));
+            }
+        }
+
+        if !locks_to_write.is_empty() {
+            self.transaction_lock
+                .batch()
+                .insert_batch(&self.transaction_lock, locks_to_write)?
+                .write()?;
+        }
+
+        Ok(())
+    }
+
+    /// Initialize a lock to None (but exists) for a given list of ObjectRefs.
+    /// If the lock already exists and is locked to a transaction, then return TransactionLockExists
+    fn initialize_locks(&self, objects: &[ObjectRef]) -> SuiResult {
+        // Use a multiget for efficiency
+        let locks = self.transaction_lock.multi_get(objects)?;
+
+        // If any locks exist and are not None, return errors for them
+        let existing_locks: Vec<ObjectRef> = locks
+            .iter()
+            .zip(objects)
+            .filter_map(|(lock_opt, objref)| {
+                if let Some(Some(_tx_digest)) = lock_opt {
+                    Some(*objref)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        if !existing_locks.is_empty() {
+            return Err(SuiError::TransactionLockExists {
+                refs: existing_locks,
+            });
+        }
+
+        // Insert where locks don't exist already
+        let refs_to_insert = locks.iter().zip(objects).filter_map(|(lock_opt, objref)| {
+            if lock_opt.is_none() {
+                Some((objref, None))
+            } else {
+                None
+            }
+        });
+        self.transaction_lock.multi_insert(refs_to_insert)?;
+
+        Ok(())
+    }
+
+    /// Removes locks for a given list of ObjectRefs.
+    fn delete_locks(&self, objects: &[ObjectRef]) -> SuiResult {
+        self.transaction_lock.multi_remove(objects)?;
+        Ok(())
+    }
+
+    /// Loop to continuously process mutating commands in a single thread from async senders
+    fn command_loop(&self, receiver: Receiver<LockServiceCommands>) {
+        info!("LockService command processing loop started");
+        while let Ok(msg) = receiver.recv() {
+            match msg {
+                LockServiceCommands::Acquire {
+                    refs,
+                    tx_digest,
+                    resp,
+                } => {
+                    let res = self.acquire_locks(&refs, tx_digest);
+                    resp.send(res).expect("Could not respond to sender!");
+                }
+                LockServiceCommands::Initialize { refs, resp } => {
+                    resp.send(self.initialize_locks(&refs))
+                        .expect("Could not respond to sender!");
```

Comment:
I would not panic here (and other places in this loop) if the oneshot channel is closed and we cannot respond. This may be due to the request from the client being dropped. I this we can just ignore the error, or just log it.